src/lib/core/GroupedCallbackList.h

```diff
@@ -0,0 +1,246 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPCallback.h>
+#include <lib/support/CodeUtils.h>
+
+#include <utility>
+
+namespace chip {
+namespace Callback {
+
+namespace detail {
+// Internal helper functions
+template <size_t Index>
+void TaggedDequeueGroup(Cancelable * cancelable);
+void EnqueueWithGroup(Cancelable * cancelable, Cancelable *& group, Cancelable * anchor, void (*cancelFn)(Cancelable *));
+void LinkGroup(Cancelable * prev, Cancelable * next);
+inline Cancelable * ClearCancelable(Cancelable * cancelable);
+} // namespace detail
+
+/**
+ * A GroupedCallbackList manages a list of Callback objects (see CHIPCallback.h).
+ * The state of the list is maintained using the prev/next pointers of each Callback.
+ * Unlike a normal linked list where entries are managed individually, this class
+ * manages a number of related callbacks as a group, with the callback function types
+ * given as template parameters.
+ *
+ * For example, a `GroupedCallbackList<SuccessFn, FailureFn>` manages groups of a
+ * `Callback<SuccessFn>` and a `Callback<FailureFn>`.
+ *
+ * Groups of callbacks are enqueued and dequeued (or cancelled) as a unit.
+ * Within a group each callback is optional (i.e. can be null), however attempting
+ * to enqueue a group where all callbacks are null has no effect.
+ */
+template <typename... T>
+class GroupedCallbackList : protected Cancelable
```

Comment:
Could you explain the reason of having this vs some `IntrusiveList<S>` where S in this case would be a struct containing 3 callbacks for our use case?

I see it tries hard to make sure list can be sparse, however given that when it is not sparse we pay the cost of 2 extra pointers (prev/next) for every entry, I am unsure if that is that much worth it.

Reply:
I think it would be useful to describe that GroupedCallbackList makes use of the prev/next pointers already existing in Callback, hence it has very limited extra overhead to a regular callback (just itself as the anchor).

That would explain why use this instead of some struct.

Reply:
This is really to avoid changing the existing OperationalSessionSetup API